### PR TITLE
tools: support ubuntu 24.04

### PR DESCRIPTION
**Pull request by mrpollo**
Updates **ubuntu.sh** to work with ubuntu 24.04, and also adds a px4-dev container, which can be used to build px4 and test via GitHub workflows in CI.

- ubuntu 24.04 is now the only supported distribution
  - this is important because we previously claimed support for older distributions, but we were not actively monitoring nor testing for support, essentially setting wrong expectations
- Introduces a new in-three container **px4-dev** which takes advantage of the updated ubuntu script
  - this container will be used to test the **ubuntu.sh** script is always up to date and working.

TODO
- [ ] github actions workflows
- [ ] check if gcc can be updated to a version available via package manager (apt)

How to use

```
# From project root
./Tools/container/docker_build.sh
./Tools/container/docker_run.sh make px4_fmu-v6x
```

also see #23745 #19982 #21808

### CODE REVIEW

File: Tools/container/docker_run.sh [hunk 1]
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
+
+if [[ -z "${DOCKER_TAG}" ]]; then
+  # The default tag name should be hardcoded
+  # to whatever we are currently using in CI on this branch
+  TAG_NAME="latest"
```
Comment (mrpollo) [inline Tools/container/docker_run.sh:6]: we need to have a test push up there

**Comment by hamishwillee** [comment]
@mrpollo Let me know at what point I can update the docs to indicate support for this build.

**Comment by mrpollo** [comment]
Once I'm done I'll let you know. For now what you should know is that we will only support the latest LTS, currently 24.04.

**Comment by hamishwillee** [comment]
> Once I'm done I'll let you know. For now what you should know is that we will only support the latest LTS, currently 24.04.

At some point that means ROS (1) and Gazebo Classic will _have_ to be dropped. Fine by me, but that will be quite the cleanup, and IMO we are not quite ready there for Gazebo. And Jae will not like us losing ROS noetic probably.

File: Tools/setup/ubuntu.sh [hunk 2]
```diff
@@ -163,29 +170,51 @@ if [[ $INSTALL_NUTTX == "true" ]]; then
...
 
 	else
-		echo "Installing arm-none-eabi-gcc-${NUTTX_GCC_VERSION}";
-		wget -O /tmp/gcc-arm-none-eabi-${NUTTX_GCC_VERSION}-linux.tar.bz2 https://armkeil.blob.core.windows.net/developer/Files/downloads/gnu-rm/${NUTTX_GCC_VERSION_SHORT}/gcc-arm-none-eabi-${NUTTX_GCC_VERSION}-${INSTALL_ARCH}-linux.tar.bz2 && \
+		# echo "Installing arm-none-eabi-gcc-${NUTTX_GCC_VERSION}";
+		echo "📌 The arm cross compiler was not found";
+		echo " * Installing arm-none-eabi-gcc-${NUTTX_GCC_VERSION}";
+		COMPILER_NAME="gcc-arm-none-eabi-${NUTTX_GCC_VERSION}"
+		COMPILER_PATH="/tmp/$COMPILER_NAME-linux.tar.bz2"
+		wget -O /tmp/gcc-arm-none-eabi-${NUTTX_GCC_VERSION}-linux.tar.bz2 https://armkeil.blob.core.windows.net/developer/Files/downloads/gnu-rm/${NUTTX_GCC_VERSION_SHORT}/gcc-arm-none-eabi-${NUTTX_GCC_VERSION}-${INSTALL_ARCH}-linux.tar.bz2
```
Comment (julianoes) [inline Tools/setup/ubuntu.sh:196]: We have to decide whether we want the arm-none-eabi toolchain shipped with 24.04, or the one downloaded manually. Right now you have both here.

File: Tools/setup/ubuntu.sh [hunk 3]
```diff
@@ -104,56 +104,63 @@ sudo DEBIAN_FRONTEND=noninteractive apt-get -y --quiet --no-install-recommends i
...
+		libncurses-dev \
+		libncursesw6 \
 		libtool \
 		pkg-config \
 		screen \
 		texinfo \
 		u-boot-tools \
 		util-linux \
 		vim-common \
+		g++-arm-linux-gnueabihf \
```
Comment (dagar) [inline Tools/setup/ubuntu.sh:159]: probably shouldn't be default

```suggestion
```

File: Tools/setup/ubuntu.sh [hunk 4]
```diff
@@ -104,56 +104,63 @@ sudo DEBIAN_FRONTEND=noninteractive apt-get -y --quiet --no-install-recommends i
...
+		gcc-arm-linux-gnueabihf \
+		g++-aarch64-linux-gnu \
```
Comment (dagar) [inline Tools/setup/ubuntu.sh:161]: probably shouldn't be default

```suggestion
		g++-aarch64-linux-gnu \
```

File: Tools/setup/ubuntu.sh [hunk 5]
```diff
@@ -2,18 +2,20 @@
 
 set -e
 
-## Bash script to setup PX4 development environment on Ubuntu LTS (22.04, 20.04, 18.04).
+## Bash script to setup PX4 development environment on Ubuntu LTS (24.04).
```
[the remark this answers is not in the document]
Comment (hamishwillee) [reply Tools/setup/ubuntu.sh:5]: Isn't it just. 
I don't particularly mind for developers. More worried about people using companion boards that rely on fixed OS distributions based on Ubuntu 22.04 or earlier. They are then screwed if we break.

[the remark this answers is not in the document]
Comment (julianoes) [reply Tools/setup/ubuntu.sh:5]: I think actually we should have all dependencies in pyenv and a local folder and not really depend on anything from the OS at all, except for simulation dependencies and probably cmake.

**Comment by slgrobotics** [comment]
Just FYI - tested, working fine.

I moved my development environment from Ubuntu 22.04 to 24.04 (different Intel I7 desktop machine). Everything seems to be working fine, including Gazebo Harmonic and _arm64_ cross-builds to Raspberry Pi 4. Coexists with ROS Jazzy (used for other unrelated project).

**Comment by mrpollo** [comment]
Thanks @slgrobotics we merged a version of this in main!

**Comment by slgrobotics** [comment]
I forgot to mention that I was testing "_main_" branch, so it was an already merged PR.

**Pull request closed without merging**